### PR TITLE
feat: aws integration

### PR DESCRIPTION
**Pull request by dagjaneiro**
Adds support for AWS S3 integration.

The integration requires an [Access Key from an IAM user](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_access-keys.html) with read/write permissions to the bucket.

The form mechanism used for WebDAV was generalised to support other integrations and used to implement this feature.

Closes standardnotes/bounties#32

---

# Integration Guide

This guide will help you set up your AWS account with an S3 bucket and User to be used by the Standard Notes FileSafe extension.

### 1. Log In

Start by logging into your [AWS account](https://console.aws.amazon.com). Make sure to pick your preferred region in the top right corner.

image

Check the [list of regions](https://docs.aws.amazon.com/general/latest/gr/rande.html) and take note of the region code you picked. In the example above I picked **Paris** so my region code is **eu-west-3**.

### 2. Create the S3 bucket

In the Services dropdown of the AWS console pick the **S3 storage service** and click on the **Create bucket** button.

Choose a name for your bucket and select the region you chose before.

image

You can **skip directly to step 3** if you don't want any additional features such as _versioning_ or _logging_ for your bucket.

In **step 3** make sure to keep ***Block all public access*** selected.

After confirming your settings you should see your new bucket.

image

### Create the IAM user with the required permissions

We'll start by creating the read/write policy for the new bucket, then, we'll create a group with that policy and finally create our user and assign it to our group.

### Create the policy

In the services dropdown select **IAM** and go to **Policies.** Click on the **Create policy** button and you should see the following screen:

image

Now click on the JSON tab and add the following policy configuration:

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "s3:ListBucket"
            ],
            "Resource": [
                "arn:aws:s3:::<bucket-name>"
            ]
        },
        {
            "Effect": "Allow",
            "Action": [
                "s3:PutObject",
                "s3:GetObject",
                "s3:DeleteObject"
            ],
            "Resource": [
                "arn:aws:s3:::<bucket-name>/*"
            ]
        }
    ]
}
```

Make sure to replace **`<bucket-name>`** with the name of the bucket you created in the previous step.

Click on the **Review policy** button, pick the name for your policy and create the policy.

image

### Create the group

Back on the **IAM** console pick **Groups** from the side menu and click on **Create New Group.**
Choose your group name, click N**ext Step** and pick the policy you created previously.

image

image

After reviewing your configuration create the group.

image

### Creating the user

Back on the **IAM** console pick **Users** from the side menu and click on **Add user**.

Choose a user name and make sure to select **Programmatic access.**

image

On the next screen add our user to the group we just created.

image

You can skip the tags screen and create the user. 

In the success screen make sure to either download the **CSV** or copy the **Access key ID** and **Secret access key** as you won't be able to view the secret access key in the future.

image

## Standard Notes

In **Standard Notes** pick **Add New** from the **Integrations** section in **FileSafe**.

image

In the **Link Integrations** page pick the **AWS S3** option and fill all the required information.

image

image

Copy the code generated to **Standard Notes** and you should see the **AWS S3** integration in the app.

image

### Testing the integration

Simply attach a file to your note to test the integration. The file should be available in **Standard Notes** and in your **S3 bucket** under **FileSafe**:

image

image

Congratulations **FileSafe** is now integrated with your **S3 bucket**.

### CODE REVIEW

File: app/lib/integrations/aws_s3_integration.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
+require 'aws-sdk-s3'
+
+class AwsS3Integration
+  def initialize(params = {})
+    if params[:authorization]
+      @auth_params = JSON.parse(Base64.decode64(params[:authorization]))
+
+      Aws.config.update({
```
Comment (mobitar) [inline app/lib/integrations/aws_s3_integration.rb]: Hmm, it updates credentials on a global object? Is there no way to do this on a per-instance basis? A global configuration can lead to a race condition if two requests come in at the same time on the same application instance, no?

Comment (mobitar) [reply app/lib/integrations/aws_s3_integration.rb]: According to the [readme](https://github.com/aws/aws-sdk-ruby) it should be possible to set credentials per object instance:

```
ec2 = Aws::EC2::Client.new(region:'us-west-2', credentials: credentials)
```

Comment (dagjaneiro) [reply app/lib/integrations/aws_s3_integration.rb]: Ah, you're right. The last ruby codebase I worked with used [unicorn](https://en.wikipedia.org/wiki/Unicorn_(web_server)) as the web server so it didn't require the code to be threadsafe. Puma does require it though.

I'll look into it 👍

**Comment by dagjaneiro** [comment]
@mobitar fixed the concurrency issue.

**Pull request merged**